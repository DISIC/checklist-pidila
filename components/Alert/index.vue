<template>
  <div :class="alertType" v-html="data.rawHTMLContent"/>
</template>

<script>
  export default {
    name: 'Alert',
    props: {
      data: {
        type: Object,
        default: () => {
          return {
            id: 'alert_000',
            type: 'info',
            rawHTMLContent: 'Contenu manquant'
          }
        }
      }
    },
    computed: {
      alertType: function () {
        return `alert alert-${this.data.type}`
      }
    }
  }
</script>

<style scoped>

</style>
