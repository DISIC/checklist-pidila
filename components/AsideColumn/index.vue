<template>
  <div class="col-aside">
    <aside>
      <slot/>
    </aside>
  </div>
</template>

<script>
  export default {
    name: 'AsideColumn'
  }
</script>
