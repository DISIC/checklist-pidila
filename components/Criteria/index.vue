<template>
  <li class="crit-item">
    <h2 class="js-expandmore crit-item-header">
      <button
        type="button"
        class="expandmore__button js-expandmore-button"
        @click="expandContent"
        :aria-expanded="isExpanded">
        <span class="crit-item-id">{{data.id}}</span>
        <span class="crit-item-title">{{data.title}}</span>
        <span class="expand-symbol" aria-hidden="true"></span>
      </button>
    </h2>
    <div class="crit-item-content js-to_expand" :hidden="isHidden">
      <div class="crit-item-test">
        <h3 class="sr-only">Tests</h3>
        <div class="crit-item-test-content" v-html="data.rawHTMLContent"></div>
      </div>

      <h3 class="sr-only">Filtres correspondants</h3>
      <ul class="tags-list">
        <li :class="cssClassForFilterGroup(filter.title)" v-for="filter in data.filters">
          <h4 class="sr-only">{{filter.title}}</h4>
          <ul class="tags-sublist">
            <li class="tags-sublist-item" v-for="tag in filter.tags">
              <span class="tag">{{tag}}</span>
            </li>
          </ul>
        </li>
      </ul>
    </div>
  </li>
</template>

<script>
  import Events from '../../bus/events'
  import FilterClasses from '../../data/filter.config'

  export default {
    name: 'Criteria',
    props: {
      data: {
        type: Object,
        default: () => {
          return {
            id: 'pi_000',
            title: 'Titre manquant',
            rawHTMLContent: 'Contenu manquant',
            filters: []
          }
        }
      }
    },
    data: () => {
      return {
        isHidden: true
      }
    },
    methods: {
      hideContent: function () {
        this.isHidden = true
      },
      showContent: function () {
        this.isHidden = false
      },
      expandContent: function () {
        this.isHidden = !this.isHidden
      },
      cssClassForFilterGroup: function (title) {
        return FilterClasses.cssClasses[title] ? FilterClasses.cssClasses[title] : 'default-labels-group'
      }
    },
    computed: {
      isExpanded: function () {
        return `${!this.isHidden}`
      }
    },
    mounted: function () {
      this.$bus.$on(Events.COLLAPSE_CRITERIA, this.hideContent)
      this.$bus.$on(Events.EXPAND_CRITERIA, this.showContent)
    }
  }
</script>
