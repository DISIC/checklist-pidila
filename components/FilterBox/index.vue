<template>
  <div class="filter-boxes">
    <h2>
      <svg class="svg-icon icon-filter" aria-hidden="true" focusable="false">
        <use xlink:href="svg/icon-sprite.svg#filter"></use>
      </svg>
      Filtrer la liste
    </h2>

    <form id="filters" class="filters nav-box-content">
      <p class="filter-boxes-info">La sélection de filtres met à jour la liste des critères sans rechargement de page.</p>

      <filter-category v-for="filter in data" :key="filter.title" :data="filter"/>

    </form>
  </div>
</template>

<script>
  import AnchorLink from '../AnchorLink/index'

  import FilterCategory from '../FilterCategory/index'

  export default {
    components: {AnchorLink, FilterCategory},
    name: 'FilterBox',
    props: {
      data: {
        type: Array,
        default: () => {
          return [
            {title: 'Filter is missing', tags: []}
          ]
        }
      }
    }
  }
</script>
