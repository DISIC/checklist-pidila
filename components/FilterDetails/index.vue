<template>
  <p class="filters-recap alert alert-info" v-html="displayFiltersRecap" aria-live="assertive" aria-atomic="true" />
</template>
<script>
  import FilterConfig from '../../data/filter.config'

  const DEFAULT_OPERATOR = ''
  const TAG_CONNECTOR = ' ou '
  const NO_FILTER_MESSAGE = 'sans aucun filtre ni recherche appliqués pour le moment.'

  export default {
    name: 'FilterDetails',
    computed: {
      numberOfCriteriasDisplayed () {
        return this.$store.getters.displayedCriteriasNumber
      },
      displayActiveFilters () {
        const orderedFilters = this.$store.getters.orderedFilters
        const numberOfCheckedTags = orderedFilters.reduce((sumOfTags, filter) => sumOfTags + filter.tags.length, 0)

        if (numberOfCheckedTags === 0) {
          return ''
        }

        return orderedFilters.reduce((displayedMessage, currentFilter) => {
          if (currentFilter.tags.length === 0) {
            return displayedMessage
          }

          const operator = FilterConfig.operators[currentFilter.title] || DEFAULT_OPERATOR
          const tagString = currentFilter.tags
            .map(tag => '<b>' + tag + '</b>')
            .join(TAG_CONNECTOR)

          return [displayedMessage, operator, tagString].join(' ')
        }, '').trim()
      },
      displayCriteriaText () {
        const criteriaTrad = {
          singular: 'critère',
          plural: 'critères'
        }
        let displayedMessage = ''
        if (this.numberOfCriteriasDisplayed <= 1) {
          displayedMessage = criteriaTrad.singular
        } else {
          displayedMessage = criteriaTrad.plural
        }
        return displayedMessage
      },
      displaySearchQuery () {
        const searchQuery = this.$store.state.searchQuery
        return searchQuery.length > 0 ? `correspondant à la recherche « <b>${searchQuery}</b> »` : ''
      },
      displayFiltersRecap () {
        let filterMessage = NO_FILTER_MESSAGE
        if (this.displaySearchQuery !== '' || this.displayActiveFilters !== '') {
          filterMessage = this.displaySearchQuery + ' ' + this.displayActiveFilters + ' <a href="./" class="filters-reset">Annuler tous les filtres</a>'
        }
        return '<span class="row-count">' + this.numberOfCriteriasDisplayed + '</span> ' +
          this.displayCriteriaText + ' ' + filterMessage
      }
    }
  }

</script>
