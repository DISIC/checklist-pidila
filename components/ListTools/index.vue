<template>
  <ul class="list-tools">
    <copy-link-button/>
    <li class="print-tool">
      <button
        class="btn btn-tool btn-printer"
        type="button"
        title="Imprimer cette liste"
        onclick="window.print()">
        <svg class="svg-icon icon-tool icon-printer" aria-hidden="true" focusable="false">
          <use xlink:href="svg/icon-sprite.svg#printer"></use>
        </svg>
        <span class="sr-only">Imprimer cette liste</span>
      </button>
    </li>
    <li class="open-tool">
      <button @click="openAllIsPressed"
        class="btn btn-tool btn-open_all"
        title="Ouvrir toute la liste des critères"
        type="button">
        <svg class="svg-icon icon-tool icon-open_all" aria-hidden="true" focusable="false">
          <use xlink:href="svg/icon-sprite.svg#open_all"></use>
        </svg>
        <span class="sr-only">Ouvrir toute la liste des critères</span>
      </button>
    </li>
    <li class="close-tool">
      <button @click="closeAllIsPressed"
        class="btn btn-tool btn-close_all"
        title="Fermer toute la liste des critères"
        type="button">
        <svg class="svg-icon icon-tool icon-close_all" aria-hidden="true" focusable="false">
          <use xlink:href="svg/icon-sprite.svg#close_all"></use>
        </svg>
        <span class="sr-only">Fermer toute la liste des critères</span>
      </button>
    </li>
  </ul>
</template>

<script>
  import Events from '../../bus/events'
  import CopyLinkButton from '~/components/CopyLinkButton'

  export default {
    name: 'ListTools',
    components: {
      CopyLinkButton
    },
    data: function () {
      return {
        isDisplayed: true,
        currentUrl: '',
        hideTooltip: true
      }
    },
    methods: {
      openAllIsPressed: function () {
        this.$bus.$emit(Events.EXPAND_CRITERIA)
      },
      closeAllIsPressed: function () {
        this.$bus.$emit(Events.COLLAPSE_CRITERIA)
      }
    }
  }
</script>
