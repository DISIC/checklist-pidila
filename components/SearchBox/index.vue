<template>
  <div class="filter-boxes">
    <h2>Rechercher dans les critères</h2>
    <form @submit.prevent="submit">
      <div class="input-group search-group">
        <input class="form-control" type="search" v-model="queryString"
               title="Rechercher dans l'ensemble des critères (exemple: accueil)" placeholder="ex. : accueil"/>
        <button type="submit" class="btn btn-search">
          <svg class="svg-icon icon-search" aria-hidden="true" focusable="false">
            <use xlink:href="svg/icon-sprite.svg#search"></use>
          </svg>
          <span class="sr-only">Valider la recherche</span>
        </button>
      </div>
    </form>
  </div>
</template>

<script>
  import Events from '~/bus/events'

  export default {
    name: 'SearchBox',
    data: () => {
      return {
        queryString: ''
      }
    },
    methods: {
      submit: function () {
        this.$store.commit('SET_SEARCH_QUERY', this.queryString)
      }
    },
    mounted () {
      this.$bus.$on(Events.STORE_INITIALIZED, () => {
        this.queryString = this.$store.state.searchQuery
      })
    }
  }
</script>
