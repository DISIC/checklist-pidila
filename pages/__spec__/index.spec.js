import {shallowMount} from '@vue/test-utils'
import flushPromises from 'flush-promises'

import Criteria from '../../components/Criteria'
import FilterBox from '../../components/FilterBox'
import FilterDetails from '../../components/FilterDetails'
import Alert from '../../components/Alert'
import ListTools from '../../components/ListTools'
import AsideColumn from '../../components/AsideColumn'
import SearchBox from '../../components/SearchBox'
import index from '../index.vue'

describe('pages | index', () => {
  const $route = {path: '/'}
  const $router = {}
  let $store = {}
  let $bus = {}
  $router.push = jest.fn()

  beforeEach(() => {
    $router.push.mockReset()
    $store = {
      state: {
        alerts: [{id: 'alert_000'}, {id: 'alert_001'}],
        globalFilters: [
          {title: '0', tags: ['0-1', '0-2']},
          {title: '1', tags: ['1-1', '1-2']}
        ],
        checkedFilters: [],
        searchQuery: ''
      },
      getters: {
        displayedCriteriasNumber: 3,
        filteredCriterias: [{title: '0'}, {title: '1'}, {title: '2'}]
      }
    }
    $store.commit = jest.fn()
    $store.dispatch = jest.fn()
    $store.dispatch.mockReturnValue(Promise.resolve())
    $bus.$emit = jest.fn()
  })

  it('should match snapshot', () => {
    // when
    const wrapper = shallowMount(index, {
      stubs: {'aside-column': AsideColumn},
      mocks: {$store, $route, $router, $bus}
    })

    // then
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should contain two Alert components', () => {
    // when
    const wrapper = shallowMount(index, {mocks: {$store, $route, $router, $bus}})

    // then
    const alertComponents = wrapper.findAll(Alert)
    expect(alertComponents).toHaveLength(2)
  })

  it('should contain one FilterDetails component', () => {
    // when
    const wrapper = shallowMount(index, {mocks: {$store, $route, $router, $bus}})

    // then
    const filterDetailsComponents = wrapper.findAll(FilterDetails)
    expect(filterDetailsComponents).toHaveLength(1)
  })

  it('should contain one ListTools component', () => {
    // when
    const wrapper = shallowMount(index, {mocks: {$store, $route, $router, $bus}})

    // then
    const listToolsComponents = wrapper.findAll(ListTools)
    expect(listToolsComponents).toHaveLength(1)
  })

  it('should contain one SearchBox component', () => {
    // when
    const wrapper = shallowMount(index, {
      stubs: {'aside-column': AsideColumn},
      mocks: {$store, $route, $router, $bus}
    })

    // then
    const searchBoxComponents = wrapper.findAll(SearchBox)
    expect(searchBoxComponents).toHaveLength(1)
  })

  it('should contain one ColumnLinks component', () => {
    // when
    const wrapper = shallowMount(index, {
      stubs: {'aside-column': AsideColumn},
      mocks: {$store, $route, $router, $bus}
    })

    // then
    const columnLinksComponents = wrapper.findAll(ColumnLinks)
    expect(columnLinksComponents).toHaveLength(1)
  })

  describe('Criterias display', () => {
    it('should display "Aucun critère ne correspond aux filtres appliqués" when no criterias are displayed', () => {
      // given
      $store.displayedCriteriasNumber = 0

      // when
      const wrapper = shallowMount(index, {mocks: {$store, $route, $router, $bus}})

      // then
      const noCriteriasMessage = wrapper.find('p.criteria-null').element.textContent.trim()
      expect(noCriteriasMessage).toBe('Aucun critère ne correspond aux filtres appliqués.')
    })

    it('should not show the criteria list when no criterias are displayed', () => {
      // given
      $store.getters.displayedCriteriasNumber = 0

      // when
      const wrapper = shallowMount(index, {mocks: {$store, $route, $router, $bus}})

      // then
      expect(wrapper.find('ul.criteria-list').element.style.display).toBe('none')
    })

    it('should show the criteria list when some criterias are displayed', () => {
      // given
      $store.getters.displayedCriteriasNumber = 2

      // when
      const wrapper = shallowMount(index, {mocks: {$store, $route, $router, $bus}})

      // then
      expect(wrapper.find('ul.criteria-list').element.style.display).not.toBe('none')
    })

    it('should show the "no criteria message" when no criterias are displayed', () => {
      // given
      $store.getters.displayedCriteriasNumber = 0

      // when
      const wrapper = shallowMount(index, {mocks: {$store, $route, $router, $bus}})

      // then
      expect(wrapper.find('p.criteria-null').element.style.display).not.toBe('none')
    })

    it('should not show the "no criteria message" when some criterias are displayed', () => {
      // given
      $store.getters.displayedCriteriasNumber = 2

      // when
      const wrapper = shallowMount(index, {mocks: {$store, $route, $router, $bus}})

      // then
      expect(wrapper.find('p.criteria-null').element.style.display).toBe('none')
    })

    it('should contain three Criteria components when store has three criterias', () => {
      // when
      const wrapper = shallowMount(index, {mocks: {$store, $route, $router, $bus}})

      // then
      const criteriaComponents = wrapper.findAll(Criteria)
      expect(criteriaComponents).toHaveLength(3)
    })

    it('should set the right props in the Criteria components', () => {
      // when
      const wrapper = shallowMount(index, {mocks: {$store, $route, $router, $bus}})

      // then
      const criteriaComponents = wrapper.findAll(Criteria)
      criteriaComponents.wrappers.forEach((component, index) => {
        expect(component.props()).toEqual({data: {title: String(index)}})
      })
    })
  })

  describe('Filter Box', () => {
    it('should contain one FilterBox component', () => {
      // when
      const wrapper = shallowMount(index, {
        stubs: {'aside-column': AsideColumn},
        mocks: {$store, $route, $router, $bus}
      })

      // then
      const filterBoxComponents = wrapper.findAll(FilterBox)
      expect(filterBoxComponents).toHaveLength(1)
    })

    it('should set the right props in the FilterBox component', () => {
      // when
      const wrapper = shallowMount(index, {
        stubs: {'aside-column': AsideColumn},
        mocks: {$store, $route, $router, $bus}
      })

      // then
      expect(wrapper.find(FilterBox).props()).toEqual({
        data: [
          {title: '0', tags: ['0-1', '0-2']},
          {title: '1', tags: ['1-1', '1-2']}
        ]
      })
    })
  })

  describe('mounted', () => {
    it('should load data from store', () => {
      // given
      shallowMount(index, {mocks: {$store, $route, $router, $bus}})

      // then
      expect($store.dispatch).toHaveBeenCalledWith('LOAD_DATA')
    })

    it('should call the "SET_FILTER_TAGS_FROM_URL_QUERY" mutation with the query params when component is mounted', async () => {
      // given
      const expectedPayload = {
        'Référentiel': ['RGAA', 'Bonnes pratiques Opquast'],
        'Profil': ['Intégration', 'Graphisme']
      }
      $route.query = {
        'Référentiel': ['RGAA', 'Bonnes pratiques Opquast'],
        'Profil': ['Intégration', 'Graphisme']
      }

      $store.commit = jest.fn()
      $store.dispatch = jest.fn()
      $store.dispatch.mockReturnValue(Promise.resolve())

      // when
      shallowMount(index, {mocks: {$store, $route, $router, $bus}})
      await flushPromises()

      // then
      expect($store.commit).toHaveBeenCalledWith('SET_FILTER_TAGS_FROM_URL_QUERY', expectedPayload)
    })

    it('should call the "SET_SEARCH_QUERY_FROM_URL_QUERY" mutation with the query params when component is mounted', async () => {
      // given
      const expectedPayload = {
        'Référentiel': ['RGAA', 'Bonnes pratiques Opquast'],
        'Profil': ['Intégration', 'Graphisme'],
        'searchQuery': 'navigation'
      }
      $route.query = {
        'Référentiel': ['RGAA', 'Bonnes pratiques Opquast'],
        'Profil': ['Intégration', 'Graphisme'],
        'searchQuery': 'navigation'
      }

      $store.commit = jest.fn()
      $store.dispatch = jest.fn()
      $store.dispatch.mockReturnValue(Promise.resolve())

      // when
      shallowMount(index, {mocks: {$store, $route, $router, $bus}})
      await flushPromises()

      // then
      expect($store.commit).toHaveBeenCalledWith('SET_SEARCH_QUERY_FROM_URL_QUERY', expectedPayload)
    })

    it('should emit the "STORE_INITIALIZED" bus event when the store initialization is completed', async () => {
      // given
      $route.query = {
        'Référentiel': ['RGAA', 'Bonnes pratiques Opquast'],
        'Profil': ['Intégration', 'Graphisme'],
        'searchQuery': 'navigation'
      }

      $store.commit = jest.fn()
      $store.dispatch = jest.fn()
      $store.dispatch.mockReturnValue(Promise.resolve())

      // when
      shallowMount(index, {mocks: {$store, $route, $router, $bus}})
      await flushPromises()

      // then
      expect($bus.$emit).toHaveBeenCalledWith('STORE_INITIALIZED')
    })
  })

  describe('URL query params management', () => {
    it('should update the URL with the right params ' +
      'when the Pages component receives one checked filter and no search query', () => {
      // given
      const wrapper = shallowMount(index, {mocks: {$store, $route, $router, $bus}})

      const payload = {path: '/', query: {'Référentiel': ['RGAA', 'Bonnes pratiques Opquast']}}
      // when
      $store.state.checkedFilters = [{title: 'Référentiel', tags: ['RGAA', 'Bonnes pratiques Opquast']}]

      // then
      expect($router.push).toHaveBeenCalledWith(payload)
    })

    it('should update the URL with the right params ' +
      'when the Pages component receives multiple checked filters and no search query', () => {
      // given
      const wrapper = shallowMount(index, {mocks: {$store, $route, $router, $bus}})

      const payload = {
        path: '/',
        query: {'Référentiel': ['RGAA', 'Bonnes pratiques Opquast'], 'Profil': ['Intégration', 'Graphisme']}
      }
      // when
      $store.state.checkedFilters = [{
        title: 'Référentiel',
        tags: ['RGAA', 'Bonnes pratiques Opquast']
      }, {title: 'Profil', tags: ['Intégration', 'Graphisme']}]

      // then
      expect($router.push).toHaveBeenCalledWith(payload)
    })

    it('should update the URL with the right params ' +
      'when the Pages component receives no checked filters and a search query', () => {
      // given
      const wrapper = shallowMount(index, {mocks: {$store, $route, $router, $bus}})
      const payload = {path: '/', query: {'searchQuery': 'navigation'}}

      // when
      $store.state.checkedFilters = []
      $store.state.searchQuery = 'navigation'

      // then
      expect($router.push).toHaveBeenCalledWith(payload)
    })

    it('should update the URL with the right params ' +
      'when the Pages component receives some checked filters and a search query', () => {
      // given
      const wrapper = shallowMount(index, {mocks: {$store, $route, $router, $bus}})
      const payload = {
        path: '/',
        query: {
          'Référentiel': ['RGAA', 'Bonnes pratiques Opquast'],
          'Profil': ['Intégration', 'Graphisme'],
          'searchQuery': 'navigation'
        }
      }

      // when
      $store.state.checkedFilters = [{
        title: 'Référentiel',
        tags: ['RGAA', 'Bonnes pratiques Opquast']
      }, {title: 'Profil', tags: ['Intégration', 'Graphisme']}]
      $store.state.searchQuery = 'navigation'

      // then
      expect($router.push).toHaveBeenCalledWith(payload)
    })
  })
})
