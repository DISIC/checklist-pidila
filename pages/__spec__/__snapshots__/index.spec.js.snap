// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages | index should match snapshot 1`] = `
<main
  class="main"
  id="main"
  role="main"
>
  <div
    class="container main-container"
  >
    <div
      class="col-group"
    >
      <div
        class="criteria-boxes col-main"
      >
        <h1>
          Liste des critères
        </h1>
         
        <alert-stub
          data="[object Object]"
        />
        <alert-stub
          data="[object Object]"
        />
         
        <filter-details-stub />
         
        <list-tools-stub />
         
        <p
          class="filters-access"
        >
          <svg
            aria-hidden="true"
            class="svg-icon icon-filter"
            focusable="false"
          >
            <use
              xlink:href="svg/icon-sprite.svg#filter"
            />
          </svg>
           
          <anchor-link-stub
            anchor="filters"
            cssclass=""
            message="Accéder aux filtres"
          />
        </p>
         
        <div>
          <ul
            class="criteria-list"
            id="criteria-list"
          >
            <criteria-stub
              data="[object Object]"
            />
            <criteria-stub
              data="[object Object]"
            />
            <criteria-stub
              data="[object Object]"
            />
          </ul>
           
          <p
            class="criteria-null"
            style="display: none;"
          >
            
            Aucun critère ne correspond aux filtres appliqués.
          
          </p>
        </div>
      </div>
       
      <div
        class="col-aside"
      >
        <aside>
          <search-box-stub />
           
          <filter-box-stub
            data="[object Object],[object Object]"
          />
           
          <column-links-stub />
        </aside>
      </div>
    </div>
  </div>
</main>
`;
